### DISCUSSION TITLE
Issue with RX Window Configuration for SX1262 on LoRaWAN Network

### DISCUSSION THREAD

**Discussion by nikorainto** [jgromes/RadioLib#1136] [Q&A]
Hello,

I am currently testing the RadioLib and LoRaWAN example on a custom board featuring an STM32L072RB MCU and a Waveshare SX1262 module. The testing environment includes a nationwide set of LoRaWAN gateways utilizing the ThingPark network platform for device management and message logging. ThingPark supports LoRaWAN v1.1.

I observed that my join request appeared in the gateway's log and a response was sent. However, it seems the RX window closes before my device can receive the join acceptance. Additionally, the JoinAccept packet log shows: "Delivery Failed Cause on RX1: Class A: LRC selected RX2(C0)." On my device I can see state being **-1116** (_RADIOLIB_LORAWAN_NO_DOWNLINK_)

![logger](https://github.com/jgromes/RadioLib/assets/26547857/a1f71de7-4d31-4a46-a5f7-151472305239)

Previously, I have been using CubeCell with LoRaWAN v1.0.2 on the same network. When the signal is weak, it can take between 10-15 seconds for the JoinAccept to appear on the device. I'm considering whether this might be the case here as well. I believe that by default, RadioLib waits 5-6 seconds for a response. Should I consider increasing `RADIOLIB_LORAWAN_JOIN_ACCEPT_DELAY_1_MS`? Does that sound correct?

You can view my current code here: [GitHub Link](https://github.com/nikorainto/radiolib-lorawan-example)

**Comment by StevenCellist** [comment]
Hi @nikorainto, the problem appears to be with ThingPark, judging by the upper screenshot as it already reports an error. The main observation is that ThingPark apparently tried to schedule at Rx1 but failed to do so with that error you copied (Class A: LRC selected RX2(C0)). 
I have never used ThingPark so I can't comment on what this means - I would recommend you do a Google search or whatever your preferred search engine would be. 
But OK, then ThingPark tried to use Rx2 at SF12 - this would be expected to work. Can you turn on RADIOLIB_DEBUG_PROTOCOL flag in your compile options and re-run your test? You would only need to do it once (or maybe twice in case it was a fluke); no need to spam the airwaves with SF12 continuously. It if not work, it not work.

**Reply by StevenCellist** [reply]
And actually, looking deeper into your screenshot (they're sooo difficult to read, please try to improve that or copy text wherever possible), it looks like there is a CUSTOM RF region used. What would this imply? What are the modifications?

**Reply by nikorainto** [reply]
Thanks for reply. I found some [documentation](https://docs.thingpark.com/thingpark-enterprise/7.3/docs/user-guide/network-tools/wireless-logger/lorawan-traffic/downlink-lorawan-unicast-packets#downlink-delivery-failure-causes) from ThingPark side about this case. 

> Over-the-air delivery error causes for RX1 downlink slot (all device classes)
> LRC selects RX2
> - (C0) LRC selected RX2

 I still don't fully understand what does it indicate.

Digita in Finland is the company that have build these LoRaWAN gateways I am utilizing. Custom [RF region](https://docs.thingpark.com/thingpark-enterprise-all-in-one/configuration-usage/rf-regions#eu-863-870-mhz-ism-band-eu868) probably indicates that 🤔 There is also something mentioned about RX1/RX2 settings:

```
RX1DROffset = 2.
RX2 Frequency = 869.525Mhz.
RX2 data rate = DR3 (SF9)
```

There is only SWD port on my custom board so I'm not sure if enabling RADIOLIB_DEBUG_PROTOCOL flag will give any debug information.

**Reply by StevenCellist** [reply]
DEBUG_PROTOCOL will be able to generate output just like normal Serial output as long as the appropriate DEBUG_PORT is configured.

What would LRC mean? Haven't heard of that abbreviation before.

**Reply by nikorainto** [reply]
Thank you for pointing this out. Until now, I've relied solely on breakpoints for debugging because I thought it was challenging to print anything using only the SWD port with the ST-Link V2 debugger.

Regarding the LRC, I found the following information:
> In the context of LoRaWAN, LRC stands for LoRaWAN Network Server (LNS) / LoRa Cloud Router (LRC), commonly referred to as just the LoRaWAN Network Server.

**Reply by StevenCellist** [reply]
To quote Mr Jan:
- Knock knock
- Who's th..
- Race condition

I was just looking at the LRC definition but you slightly beat me to it. 

Actually, I am not completely sure whether you can get any output from the SWD port, maybe @jgromes can advise on this. I thought that given the -1116 error code, you were able to generate Serial output, and if you can get that, then the DEBUG mode will be able to work just the same.
The problem is twofold: I don't know ThingPark and have no idea what the details are on their stack, nor do we know anything about how your device behaves. So please try to get as many details as possible using whatever verbose or debug mode you can get working.. 
A good first step would be to see if your device works with TTN, because then we know the problems are not with the device and rather with something in ThingPark - which might still be difficult to debug, but at least we know where to look!

**Reply by jgromes** [reply]
@nikorainto not sure there is a way to easmlessly output the current UART-based debug output via the STlink, however, as a workaround you can put a breakpoint into `Module::serialPrintf` and examine its arguments.

**Reply by nikorainto** [reply]
Thank you both for input! I will next try to verify first with TTN that common things should work and once that is complete I will come back to debug with this 3rd party service stack 👍

**Reply by nikorainto** [reply]
With TTN everything appears to work flawlessly, so I'll continue testing with it for the time being

![image](https://github.com/jgromes/RadioLib/assets/26547857/cddd3e94-e413-4aed-a55f-4d4dc5f405d6)

**Reply by HeadBoffin** [reply]
> With TTN everything appears to work flawlessly

FYI, that's the platform we use as our primary test, with ChirpStack as the secondary but I'll add "Murata-style" device to the list of things to try.


> Should I consider increasing RADIOLIB_LORAWAN_JOIN_ACCEPT_DELAY_1_MS? Does that sound correct?

No. The Join parameters are part of the spec. The thing to look at hardest initially is the very marginal signal that is being heard by the gateway - there is usually some asymmetry between devices & gateways, the former having small bits of wire and the latter having nice chunky well positioned installations.

**Reply by nikorainto** [reply]
I now face a different issue that is not related to TTN or ThingPark. Instead, it's related to my coding skills and my current difficulty in debugging the code effectively. This is the current code I'm testing with to have both BME280 sensor and Radiolib working on same project

```
#include <Wire.h>
#include <Adafruit_Sensor.h>
#include <Adafruit_BME280.h>
#include "config.h"

Adafruit_BME280 bme;

void setup()
{
  Wire.begin();
  bme.begin(0x76);

  radio.begin();
  radio.setRfSwitchPins(PA3, PA2);

  node.beginOTAA(joinEUI, devEUI, nwkKey, appKey);
  node.activateOTAA();
}

void loop()
{
}
```
For some reason I'm not able to understand when both `Wire.begin();` and `node.activateOTAA();` are enabled then code will go to infinite loop without obvious reason. If either is commented then code works normally. Does anyone with more knowledge than me understand what is happening even without having debug logs?

**Reply by jgromes** [reply]
Nothing obvious comes to mind, though since you have an SWD, you should be able to just run the code with a debugger and then stop execution once it gets into that infinite loop ... then you can just examine the stack trace to tell where that loops is, and how the program got there.

**Reply by nikorainto** [reply]
I tried to debug this more but seems like dead end. Code does not stop to any breakpoint in any line but instead starts from this
```
Reset_Handler:  
   ldr   r0, =_estack
   mov   sp, r0          /* set stack pointer */
```
and when clicking pause it stops to this line
```
Infinite_Loop:
  b  Infinite_Loop
```
with following call stack
```
WWDG_IRQHandler@0x0800c52c (c:\Users\me\.platformio\packages\framework-arduinoststm32\system\Drivers\CMSIS\Device\ST\STM32L0xx\Source\Templates\gcc\startup_stm32l072xx.s:110)
<signal handler called>@0xfffffff9 (Unknown Source:0)
__libc_init_array@0x08010cc0 (\__libc_init_array.dbgasm:1)
Reset_Handler@0x0800c50e (c:\Users\me\.platformio\packages\framework-arduinoststm32\system\Drivers\CMSIS\Device\ST\STM32L0xx\Source\Templates\gcc\startup_stm32l072xx.s:89)
```
and that keeps repeating if I try continue and pause again 🤔Also nothing special can be seen in terminal debug console. If I run code with just Radiolib parts or only BME280 parts then everything works

**Reply by jgromes** [reply]
This would deserve a more in-depth exploration with the debugger, but for what it's worth ...

1. On STM32, WWDG is the windowed watchdog, and if I remember correctly, the default implementation of its IRQ handler is an infinite loop. So if the watchdog is being enabled and then timing out (e.g. if some fucntion takes too long without feeding the watchdog), it will trigger the IRQ, hence the infinite loop.
2. However, I also seem to recall that the `WWDG_IRQHandler` is the point in the interrupt vector that interrupts will jump to if the don't have an IRQ handler. So if there's an interrupt that's missing its handler, it would jump into `WWDG_IRQHandler` without actually being triggered by the WWDG.

Take the above with a grain of salt, it's been a while since I've seen this behavior of STM32 program ending up in `WWDG_IRQHandler`

**Reply by nikorainto** [reply]
Reason for this was found. I had wrong openocd_target defined on my board definition. Had to be **stm32l0_dual_bank** instead of **stm32l0**
```
  "debug": {
    "jlink_device": "STM32L072RB",
    "openocd_target": "stm32l0_dual_bank",
    "svd_path": "STM32L07x.svd"
  },
```

**Reply by nikorainto** [reply]
So I got everything working on TTN on a self hosted gateway and getting back to this issue. I found documentation from service provider which states following:

- The device must comply with the LoRaWAN specification (at least 1.0).
- The device must support the EU868 frequency band, at least the three mandatory frequencies of the EU868 band (the rest of the frequencies are communicated to the device either during the Join process (OTAA) or to ABP devices via separate MAC commands).
- The device must have a unique DevEUI identifier, and for security purposes, also a unique encryption key.
- The device must adhere to the duty cycle restrictions as defined by the communications authority, which is typically 1%.
- The device must support the ADR feature (Adaptive Data Rate) and, in general, the message type should be unconfirmed.

Other considerations:

- The spreading factor used in the Rx2 window is only the initial value used in the OTAA device Join procedure and in the first messages of ABP devices. The device must be capable of changing this data rate of the listening window according to what the network commands it, either in the OTAA Join or to ABP devices via MAC commands.
- By default, our network uses a downlink data rate offset value of +2.
- By default, the delay for the RX1 window is 1000ms and for the RX2 window 2000ms. In the Join procedure, the corresponding delays are 5000ms and 6000ms.
- The LoRaWAN channels used by Digita are: (initially, it is most important for the device to understand to use the mandatory channels LC1-LC3 specified by the EU868 spec and the Digita network's RX2 frequency; other channels are communicated to the device via MAC commands):

LC1 868.1 MHz
LC2 868.3 MHz
LC3 868.5 MHz
RX2 869.525 MHz

So as far as I understand all these should be OK right?

Device "model" I selected from ThingPark/Digita portal is: **LoRaWAN 1.1 revA - class A  eu868, eu433, cn779, kr920, in865, ru864** and there is no specific selections for **Frequency plan** and **Regional parameters version** like in TTN portal

**Reply by StevenCellist** [reply]
Looks about right to me! Weird that it doesn't select a specific plan (e.g. EU868) but the Regional Parameters are included in the *1.1 revA*.

Only thing that looks a bit odd to me is the downlink datarate offset of 2 - not exactly sure how that would play out, i.e. if network communicates that by itself or if additional configuration is required.

**Reply by ChangYanChu** [reply]
@nikorainto  Hi , I also encountered the same issue as you, buddy. Regarding what you mentioned: "By default, the delay for the RX1 window is 1000ms and for the RX2 window 2000ms. In the Join procedure, the corresponding delays are 5000ms and 6000ms." Did you later manage to establish the registration by modifying the default RX1 window and RX2 window?

**Reply by HeadBoffin** [reply]
The Join Rx delays are set in the stone in the specification. When you join, the LNS will tell you what the Rx1 delay is for that server. So none of this should be necessary.

**Reply by qbsho** [reply]
Did you get it working?

**Comment by K4KDR** [comment]
Don't have a specific answer for you, but as I've been learning how to use the various LoRaWAN capabilities in RadioLib, a LOT of useful information has been revealed by un-commenting one or more of the 'debug' lines in **/src/BuildOptUser.h**

![Screenshot from 2024-06-26 17-35-34](https://github.com/jgromes/RadioLib/assets/19886368/acb1fc66-57a0-4d89-a6d5-c5440bb62e31)

... besides revealing all kinds of LoRaWAN info to point you in the right direction, the debug info that appears in your serial terminal can of course also be a huge help with a module like the BME280, too!

**Reply by nikorainto** [reply]
There is just an issue that I have only SWD port available on my board and no serial port so I think that option will not work for me 😅

**Reply by K4KDR** [reply]
I only have a couple of STM32 devices but all of them allow me to connect via a terminal app (using CuteCom on linux) to the device that usually presents itself as something like **/dev/ttyACM0**.    But I'm very much a novice with these modules so perhaps that is not true for all of those boards

**Reply by nikorainto** [reply]
Yes, I also have some Arduino and STM32 development boards without the LoRa module that work well with serial port debugging. However, I am new to this field also and recently built a custom board with only an SWD port. I thought it would provide the same debugging capabilities, but it seems that I can only rely on breakpoints in the code for debugging 😬

**Reply by nikorainto** [reply]
I have a separate project for reading BME280 sensor values and it works well. Also I have another with Radiolib and LoRaWAN and that works. Now I'm trying to combine these two but ran into this problem

**Reply by HeadBoffin** [reply]
> I am new to this field also and recently built a custom board with only an SWD port.

You aren't alone in making this error but the only solution is to either bodge two wires on to the side of the chip or re-do your board. Once you have done the development work you can return to your no-serial PCB for deployment. However I'd point ou that the minimum I have on production boards are either pads or 1.27mm header - the SWD connector provides a Gnd, so the minimum is ONE pin for Tx.

> Does anyone with more knowledge than me understand what is happening even without having debug logs?

Advanced guessing would be to look at an overlap of pin use. Or as Jan says, use the debugger even if it takes an evening of learning the debugger rather than hoping you can operate it by brute force. Or, as I do, use a bigger / better board for development & debugging and then port it over.
